### PR TITLE
Kube openapi for openapi spec

### PR DESCRIPTION
**Pull request by xpivarc**
**What this PR does / why we need it**:
Switching to kube-openapi generator for openapi. This should provide better openapi spec and therefore should also help with generating clients.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

This PR don't directly fix any issue. Pinging #3177 as it might be helpful.

**Special notes for your reviewer**:
Please take extra time to look at generated swagger as this also affect our validating webhooks.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-api/api.go [hunk 1]
```diff
@@ -375,10 +374,10 @@ func (app *virtAPIApp) composeSubresources() {
 
 				response.WriteAsJson(list)
 			}).
-			Operation("getAPIResources").
+			Operation("getAPISubResources").
```
Comment (davidvossel) [inline pkg/virt-api/api.go:377]: what does changing the operation name here impact?

Comment (xpivarc) [reply pkg/virt-api/api.go:377]: I believe that impact is only in the documentation. I changed it because kube-openapi demands unique operations...

**Comment by xpivarc** [comment]
> This all looks generally like what i'd expect. Realistically, it's difficult to review this in it's entirety do to all the generated files. We'll have to lean on our CI for validating this. It looks like all the admission tests in aggregate cover the openapi generation changes pretty well. Does that sound accurate? basically, is there anything else we should be doing to prove this change that doesn't already exist?

Yes admission tests are one of many. As every test which posts our crds to API results in validating against the schema. I can only think of a test where we check objects with some non-common fields (not sure if we have every field cover).

**Comment by rmohr** [comment]
> This all looks generally like what i'd expect. Realistically, it's difficult to review this in it's entirety do to all the generated files. We'll have to lean on our CI for validating this.

I also think that we don't have to review the whole generated swagger file, but let's ensure that no new `required` fields appeared, or that `required` fields became optional. That should be doable. We sometimes added `+optional` and sometimes did not. Now `openapi-gen` finally interprets them.

**Comment by davidvossel** [comment]
> We sometimes added +optional and sometimes did not. Now openapi-gen finally interprets them.

Oh, well if our previous openapi didn't interpret optional vs required, then we may very well have a subtle change in behavior occur. That explains why the datavolume.Spec.PVC had to be set to get our unit tests to pass now. It really is a required field, but we didn't actually require it previously because of how the spec was interpreted.

I believe this risk of subtle behavior changes is something we're going to have to accept. Realistically we can't have a way of knowing if the entirety of every VM/VMI/VMIRS/etc... objects in production will pass now. However, we do know that this change is the foundation we want moving forward.

**Comment by xpivarc** [comment]
Failing test point to another missed behavior. I think Timer shouldn't be `inline` or so API is telling me.

**Comment by davidvossel** [comment]
> Failing test point to another missed behavior. I think Timer shouldn't be inline or so API is telling me.

good catch!

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 2]
```diff
@@ -685,17 +696,19 @@ const (
 	VirtualMachineInstanceReplicaSetReplicaPaused VirtualMachineInstanceReplicaSetConditionType = "ReplicaPaused"
 )
 
-// ---
+//
 // +k8s:openapi-gen=true
 type VirtualMachineInstanceTemplateSpec struct {
+	// +kubebuilder:pruning:PreserveUnknownFields
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:702]: why was this added?

Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:702]: Nice catch. That was me testing which markers are detected by kube-openapi. This has no effect as far as I know. Should I remove it? And thanks for go.mod. I have a habit to not commit it.

Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go]: alrighty, i removed it.

File: staging/src/kubevirt.io/client-go/api/v1/schema.go [hunk 3]
```diff
@@ -628,22 +628,24 @@ type ClockOffsetUTC struct {
...
-// ---
+//
 // +k8s:openapi-gen=true
 type ClockOffsetTimezone string
 
 // Represents the clock and timers of a vmi.
-// ---
+//
 // +k8s:openapi-gen=true
+// +kubebuilder:pruning:PreserveUnknownFields
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/schema.go:638]: here's another seemingly random kubebuilder marker

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 4]
```diff
@@ -354,7 +360,9 @@ type VirtualMachineInstanceMigrationState struct {
 	TargetNodeDomainDetected bool `json:"targetNodeDomainDetected,omitempty"`
 	// The address of the target node to use for the migration
 	TargetNodeAddress string `json:"targetNodeAddress,omitempty"`
+	// +kubebuilder:validation:Type=object
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:363]: why is this here?

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 5]
```diff
@@ -354,7 +360,9 @@ type VirtualMachineInstanceMigrationState struct {
...
 	// The list of ports opened for live migration on the destination node
+	// +k8s:openapi-gen=false
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:365]: is there a reason openapi-gen=false here?

File: staging/src/kubevirt.io/client-go/api/v1/schema.go (hunk 3 above)
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:638]: I will remove all occurrences of this.

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 4 above)
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:363]: same....

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 5 above)
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:365]: And that's a problem to discuss (totally forgot about this). map[int]int is invalid type. It would be bad to not generate this. We can work around this by declaring OpenAPISchemaType() or we can change our code to map[string]int.

Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:365]: OH, that's super interesting. I just hit the same thing when I was messing with the generation logic myself. 

How exactly would the workaround look with OpenAPISchemaType ?

Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:365]: Actually I tried it and it didn't work... so https://github.com/kubevirt/kubevirt/pull/3375#issuecomment-624057689 if I didn't miss anything...

Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:365]: can we update this now that the ports map has been fixed? once this is removed as well as the kubebuilder markers, i think we'll be really close.

Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:365]: Yes, already working on this. Also, I rebased and found more `+k8s:openapi-gen=false ` which should be there.

**Comment by xpivarc** [comment]
@davidvossel Hi, I reworked it a little bit. There were +k8s:openapi-gen=false on metadata which I removed. Removed kubebuilder markers so there should be only things which does something actually....

File: go.mod [hunk 6]
```diff
@@ -111,3 +112,5 @@ replace (
 
 	kubevirt.io/client-go => ./staging/src/kubevirt.io/client-go
 )
+
+go 1.13
```
Comment (davidvossel) [inline go.mod]: this part of go.mod needs to be removed. it slips in occasionally

Comment (davidvossel) [reply go.mod]: i fixed this myself on this pr since it was super simple.

Comment (xpivarc) [reply go.mod]: Thanks, IDE...
